nRF Battery measurement issue

Hello guys @galak @pabigot @nashif @carlescufi @tbursztyka ,

**Describe the bug**
According to the [nRF-Battery-Voltage-Sample](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/boards/nrf/battery) it is possible to use a voltage divider (vbatt). Compiling this, with the suggested board works fine:
` zephyr/samples/boards/nrf/battery $ west build -p auto -b thingy52_nrf52832 ./ --pristine`

 But if you comment the vbat-node in the devicetree, or just set its status to `status="disabled"` there is an compiler error:

`zephyr/include/generated/devicetree_unfixed.h:1344:43: warning: initialization of 'unsigned char' from 'char *' makes integer from pointer without a cast [-Wint-conversion]
 1344 | #define DT_N_S_soc_S_adc_40007000_P_label "ADC_0" `

(see full log below)

**To Reproduce**
Steps to reproduce the behavior:
1. Go to `zephyr/samples/boards/nrf/battery`
2. west build -p auto -b thingy52_nrf52832
3. it works
4. In `zephyr/boards/arm/thingy52_nrf52832´ add status to thingy52_nrf52832.dts:
```
vbatt {
        status="disabled";
        compatible = "voltage-divider";
	io-channels = <&adc 4>;
	output-ohms = <180000>;
	full-ohms = <(1500000 + 180000)>;
	power-gpios = <&sx1509b 4 0>;
};
``` 
5. Try to build again: `west build -p auto -b thingy52_nrf52832 --pristine`
6. see the error

**Expected behavior**
I want to make a battery-voltage-measurement on the nrf52840DK/dongle and this one does not use/need a voltage divider, that is why there is no "vbatt" in their devicetree. It should work, since it is also an nRF-Chip and I did it before with other RTOS'.
Nevertheless, I think there must be an problem in the battery library, since it is not even working with the suggested board (thingy52_nrf52832).

**Logs and console output**
```
[98/155] Building C object CMakeFiles/app.dir/src/battery.c.obj
FAILED: CMakeFiles/app.dir/src/battery.c.obj 
ccache /home/smiethe/Programme/gcc-arm-none-eabi-9-2019-q4-major/bin/arm-none-eabi-gcc -DBUILD_VERSION=zephyr-v2.6.0-406-g9530d3237661 -DKERNEL -DNRF52832_XXAA -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR__=1 -I/home/smiethe/git/zephyrproject1/zephyr/include -Izephyr/include/generated -I/home/smiethe/git/zephyrproject1/zephyr/soc/arm/nordic_nrf/nrf52 -I/home/smiethe/git/zephyrproject1/modules/hal/cmsis/CMSIS/Core/Include -I/home/smiethe/git/zephyrproject1/modules/hal/nordic/nrfx -I/home/smiethe/git/zephyrproject1/modules/hal/nordic/nrfx/drivers/include -I/home/smiethe/git/zephyrproject1/modules/hal/nordic/nrfx/mdk -I/home/smiethe/git/zephyrproject1/zephyr/modules/hal_nordic/nrfx/. -isystem /home/smiethe/git/zephyrproject1/zephyr/lib/libc/minimal/include -isystem /home/smiethe/Programme/gcc-arm-none-eabi-9-2019-q4-major/bin/../lib/gcc/arm-none-eabi/9.2.1/include -isystem /home/smiethe/Programme/gcc-arm-none-eabi-9-2019-q4-major/bin/../lib/gcc/arm-none-eabi/9.2.1/include-fixed -Os -imacros /home/smiethe/git/zephyrproject1/zephyr/samples/boards/nrf/battery/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m4 -mthumb -mabi=aapcs -imacros /home/smiethe/git/zephyrproject1/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-address-of-packed-member -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/smiethe/git/zephyrproject1/zephyr/samples/boards/nrf/battery=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/smiethe/git/zephyrproject1/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/smiethe/git/zephyrproject1=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT CMakeFiles/app.dir/src/battery.c.obj -MF CMakeFiles/app.dir/src/battery.c.obj.d -o CMakeFiles/app.dir/src/battery.c.obj -c ../src/battery.c
In file included from /home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:19,
                 from /home/smiethe/git/zephyrproject1/zephyr/include/arch/arm/aarch32/arch.h:20,
                 from /home/smiethe/git/zephyrproject1/zephyr/include/arch/cpu.h:19,
                 from /home/smiethe/git/zephyrproject1/zephyr/include/kernel_includes.h:33,
                 from /home/smiethe/git/zephyrproject1/zephyr/include/kernel.h:17,
                 from /home/smiethe/git/zephyrproject1/zephyr/include/zephyr.h:18,
                 from ../src/battery.c:12:
zephyr/include/generated/devicetree_unfixed.h:1344:43: warning: initialization of 'unsigned char' from 'char *' makes integer from pointer without a cast [-Wint-conversion]
 1344 | #define DT_N_S_soc_S_adc_40007000_P_label "ADC_0"
      |                                           ^~~~~~~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:2326:24: note: in expansion of macro 'DT_N_S_soc_S_adc_40007000_P_label'
 2326 | #define DT_CAT(a1, a2) a1 ## a2
      |                        ^~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:527:32: note: in expansion of macro 'DT_CAT'
  527 | #define DT_PROP(node_id, prop) DT_CAT(node_id, _P_##prop)
      |                                ^~~~~~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:648:27: note: in expansion of macro 'DT_PROP'
  648 | #define DT_LABEL(node_id) DT_PROP(node_id, label)
      |                           ^~~~~~~
../src/battery.c:74:3: note: in expansion of macro 'DT_LABEL'
   74 |   DT_LABEL(DT_NODELABEL(adc)),
      |   ^~~~~~~~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:2326:24: note: in expansion of macro 'DT_N_NODELABEL_adc'
 2326 | #define DT_CAT(a1, a2) a1 ## a2
      |                        ^~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:173:29: note: in expansion of macro 'DT_CAT'
  173 | #define DT_NODELABEL(label) DT_CAT(DT_N_NODELABEL_, label)
      |                             ^~~~~~
../src/battery.c:74:12: note: in expansion of macro 'DT_NODELABEL'
   74 |   DT_LABEL(DT_NODELABEL(adc)),
      |            ^~~~~~~~~~~~
zephyr/include/generated/devicetree_unfixed.h:1344:43: note: (near initialization for 'divider_config.io_channel.channel')
 1344 | #define DT_N_S_soc_S_adc_40007000_P_label "ADC_0"
      |                                           ^~~~~~~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:2326:24: note: in expansion of macro 'DT_N_S_soc_S_adc_40007000_P_label'
 2326 | #define DT_CAT(a1, a2) a1 ## a2
      |                        ^~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:527:32: note: in expansion of macro 'DT_CAT'
  527 | #define DT_PROP(node_id, prop) DT_CAT(node_id, _P_##prop)
      |                                ^~~~~~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:648:27: note: in expansion of macro 'DT_PROP'
  648 | #define DT_LABEL(node_id) DT_PROP(node_id, label)
      |                           ^~~~~~~
../src/battery.c:74:3: note: in expansion of macro 'DT_LABEL'
   74 |   DT_LABEL(DT_NODELABEL(adc)),
      |   ^~~~~~~~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:2326:24: note: in expansion of macro 'DT_N_NODELABEL_adc'
 2326 | #define DT_CAT(a1, a2) a1 ## a2
      |                        ^~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:173:29: note: in expansion of macro 'DT_CAT'
  173 | #define DT_NODELABEL(label) DT_CAT(DT_N_NODELABEL_, label)
      |                             ^~~~~~
../src/battery.c:74:12: note: in expansion of macro 'DT_NODELABEL'
   74 |   DT_LABEL(DT_NODELABEL(adc)),
      |            ^~~~~~~~~~~~
zephyr/include/generated/devicetree_unfixed.h:1344:43: error: initializer element is not computable at load time
 1344 | #define DT_N_S_soc_S_adc_40007000_P_label "ADC_0"
      |                                           ^~~~~~~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:2326:24: note: in expansion of macro 'DT_N_S_soc_S_adc_40007000_P_label'
 2326 | #define DT_CAT(a1, a2) a1 ## a2
      |                        ^~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:527:32: note: in expansion of macro 'DT_CAT'
  527 | #define DT_PROP(node_id, prop) DT_CAT(node_id, _P_##prop)
      |                                ^~~~~~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:648:27: note: in expansion of macro 'DT_PROP'
  648 | #define DT_LABEL(node_id) DT_PROP(node_id, label)
      |                           ^~~~~~~
../src/battery.c:74:3: note: in expansion of macro 'DT_LABEL'
   74 |   DT_LABEL(DT_NODELABEL(adc)),
      |   ^~~~~~~~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:2326:24: note: in expansion of macro 'DT_N_NODELABEL_adc'
 2326 | #define DT_CAT(a1, a2) a1 ## a2
      |                        ^~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:173:29: note: in expansion of macro 'DT_CAT'
  173 | #define DT_NODELABEL(label) DT_CAT(DT_N_NODELABEL_, label)
      |                             ^~~~~~
../src/battery.c:74:12: note: in expansion of macro 'DT_NODELABEL'
   74 |   DT_LABEL(DT_NODELABEL(adc)),
      |            ^~~~~~~~~~~~
zephyr/include/generated/devicetree_unfixed.h:1344:43: note: (near initialization for 'divider_config.io_channel.channel')
 1344 | #define DT_N_S_soc_S_adc_40007000_P_label "ADC_0"
      |                                           ^~~~~~~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:2326:24: note: in expansion of macro 'DT_N_S_soc_S_adc_40007000_P_label'
 2326 | #define DT_CAT(a1, a2) a1 ## a2
      |                        ^~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:527:32: note: in expansion of macro 'DT_CAT'
  527 | #define DT_PROP(node_id, prop) DT_CAT(node_id, _P_##prop)
      |                                ^~~~~~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:648:27: note: in expansion of macro 'DT_PROP'
  648 | #define DT_LABEL(node_id) DT_PROP(node_id, label)
      |                           ^~~~~~~
../src/battery.c:74:3: note: in expansion of macro 'DT_LABEL'
   74 |   DT_LABEL(DT_NODELABEL(adc)),
      |   ^~~~~~~~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:2326:24: note: in expansion of macro 'DT_N_NODELABEL_adc'
 2326 | #define DT_CAT(a1, a2) a1 ## a2
      |                        ^~
/home/smiethe/git/zephyrproject1/zephyr/include/devicetree.h:173:29: note: in expansion of macro 'DT_CAT'
  173 | #define DT_NODELABEL(label) DT_CAT(DT_N_NODELABEL_, label)
      |                             ^~~~~~
../src/battery.c:74:12: note: in expansion of macro 'DT_NODELABEL'
   74 |   DT_LABEL(DT_NODELABEL(adc)),
      |            ^~~~~~~~~~~~
[107/155] Building C object zephyr/drivers/regulator/CMakeFiles/drivers__regulator.dir/regulator_fixed.c.obj
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /home/smiethe/git/zephyrproject1/zephyr/samples/boards/nrf/battery/build

```

**Environment**
 - OS: Linux/Gentoo
 - Toolchain: gcc, West, Zephyr-v2.6.0-406-g9530d3237661

Thanks for your help.
Best regards,
Sebastian

--- a/samples/boards/nrf/battery/README.rst
+++ b/samples/boards/nrf/battery/README.rst
@@ -73,7 +73,7 @@ The code can be found in :zephyr_file:`samples/boards/nrf/battery`.
 
 .. zephyr-app-commands::
    :zephyr-app: samples/boards/nrf/battery
-   :board: nrf52_pca20020
+   :board: thingy52_nrf52832
    :goals: build flash
    :compact:
 

--- a/samples/boards/nrf/battery/README.rst
+++ b/samples/boards/nrf/battery/README.rst
@@ -11,29 +11,40 @@ infrastructure to measure the voltage of the device power supply.  Two
 power supply configurations are supported:
 
 * If the board devicetree has a ``/vbatt`` node with compatible
-  ``voltage-divider`` then the voltage is measured using that divider.
-* Otherwise the power source is assumed to be directly connected to the
-  digital voltage signal, and the internal source for ``Vdd`` is
-  selected.
+  ``voltage-divider`` then the voltage is measured using that divider. An
+  example of a devicetree node describing a voltage divider for battery
+  monitoring is:
 
-An example of a devicetree node describing a voltage divider for battery
-monitoring is:
+   .. code-block:: devicetree
 
-.. code-block:: none
+      / {
+         vbatt {
+            compatible = "voltage-divider";
+            io-channels = <&adc 4>;
+            output-ohms = <180000>;
+            full-ohms = <(1500000 + 180000)>;
+            power-gpios = <&sx1509b 4 0>;
+         };
+      };
 
-   / {
-   	vbatt {
-   		compatible = "voltage-divider";
-   		io-channels = <&adc 4>;
-   		output-ohms = <180000>;
-   		full-ohms = <(1500000 + 180000)>;
-   		power-gpios = <&sx1509b 4 0>;
-   	};
-   };
+* If the board does not have a voltage divider and so no ``/vbatt`` node
+  present, the ADC configuration (device and channel) needs to be provided via
+  the ``zephyr,user`` node. In this case the power source is assumed to be
+  directly connected to the digital voltage signal, and the internal source for
+  ``Vdd`` is selected. An example of a Devicetree configuration for this case is
+  shown below:
+
+   .. code-block :: devicetree
+
+      / {
+         zephyr,user {
+            io-channels = <&adc 4>;
+         };
+      };
 
 Note that in many cases where there is no voltage divider the digital
 voltage will be fed from a regulator that provides a fixed voltage
-regardless of source voltage, rather than by the source directly.  In
+regardless of source voltage, rather than by the source directly. In
 configuration involving a regulator the measured voltage will be
 constant.
 

--- a/samples/boards/nrf/battery/src/battery.c
+++ b/samples/boards/nrf/battery/src/battery.c
@@ -21,6 +21,7 @@
 LOG_MODULE_REGISTER(BATTERY, CONFIG_ADC_LOG_LEVEL);
 
 #define VBATT DT_PATH(vbatt)
+#define ZEPHYR_USER DT_PATH(zephyr_user)
 
 #ifdef CONFIG_BOARD_THINGY52_NRF52832
 /* This board uses a divider that reduces max voltage to
@@ -71,7 +72,7 @@ static const struct divider_config divider_config = {
 	.full_ohm = DT_PROP(VBATT, full_ohms),
 #else /* /vbatt exists */
 	.io_channel = {
-		DT_LABEL(DT_NODELABEL(adc)),
+		DT_IO_CHANNELS_INPUT(ZEPHYR_USER),
 	},
 #endif /* /vbatt exists */
 };
@@ -84,7 +85,11 @@ struct divider_data {
 	int16_t raw;
 };
 static struct divider_data divider_data = {
+#if DT_NODE_HAS_STATUS(VBATT, okay)
 	.adc = DEVICE_DT_GET(DT_IO_CHANNELS_CTLR(VBATT)),
+#else
+	.adc = DEVICE_DT_GET(DT_IO_CHANNELS_CTLR(ZEPHYR_USER)),
+#endif
 };
 
 static int divider_setup(void)

